opm/autodiff/AquiferInterface.hpp

```diff
@@ -0,0 +1,224 @@
+    typedef typename GET_PROP_TYPE(TypeTag, FluidSystem) FluidSystem;
+    typedef typename GET_PROP_TYPE(TypeTag, Indices) BlackoilIndices;
+    typedef typename GET_PROP_TYPE(TypeTag, RateVector) RateVector;
+    typedef typename GET_PROP_TYPE(TypeTag, IntensiveQuantities) IntensiveQuantities;
+
+    enum { enableTemperature = GET_PROP_VALUE(TypeTag, EnableTemperature) };
+    enum { enableEnergy = GET_PROP_VALUE(TypeTag, EnableEnergy) };
+
+    static const int numEq = BlackoilIndices::numEq;
+    typedef double Scalar;
+
+    typedef DenseAd::Evaluation<double, /*size=*/numEq> Eval;
+
+    typedef Opm::BlackOilFluidState<Eval, FluidSystem, enableTemperature, enableEnergy, BlackoilIndices::gasEnabled, BlackoilIndices::numPhases> FluidState;
+
+    static const auto waterCompIdx = FluidSystem::waterCompIdx;
+    static const auto waterPhaseIdx = FluidSystem::waterPhaseIdx;
+
+    // Constructor
+    AquiferInterface( const Aquancon::AquanconOutput& connection,
+                      const std::unordered_map<int, int>& cartesian_to_compressed,
+                      const Simulator& ebosSimulator)
+                    : connection_(connection)
+                    , ebos_simulator_(ebosSimulator)
+                    , cartesian_to_compressed_(cartesian_to_compressed)
+    {}
+
+    // Deconstructor
+    virtual ~AquiferInterface() {}
+
+    void initialSolutionApplied()
+    {
+      initQuantities(connection_);
+    }
+
+    void beginTimeStep()
+    {
+      ElementContext elemCtx(ebos_simulator_);
+      auto elemIt = ebos_simulator_.gridView().template begin<0>();
+      const auto& elemEndIt = ebos_simulator_.gridView().template end<0>();
+      for (; elemIt != elemEndIt; ++elemIt) {
+        const auto& elem = *elemIt;
+
+        elemCtx.updatePrimaryStencil(elem);
+
+        int cellIdx = elemCtx.globalSpaceIndex(0, 0);
+        int idx = cellToConnectionIdx_[cellIdx];
+        if (idx < 0)
+        continue;
+
+        elemCtx.updateIntensiveQuantities(0);
+        const auto& iq = elemCtx.intensiveQuantities(0, 0);
+        pressure_previous_[idx] = Opm::getValue(iq.fluidState().pressure(waterPhaseIdx));
+      }
+    }
+
+    template <class Context>
+    void addToSource(RateVector& rates, const Context& context, unsigned spaceIdx, unsigned timeIdx)
+    {
+      unsigned cellIdx = context.globalSpaceIndex(spaceIdx, timeIdx);
+
+      int idx = cellToConnectionIdx_[cellIdx];
+      if (idx < 0)
+      return;
+
+      // We are dereferencing the value of IntensiveQuantities because cachedIntensiveQuantities return a const pointer to
+      // IntensiveQuantities of that particular cell_id
+      const IntensiveQuantities intQuants = context.intensiveQuantities(spaceIdx, timeIdx);
+      // This is the pressure at td + dt
+      updateCellPressure(pressure_current_,idx,intQuants);
+      updateCellDensity(idx,intQuants);
+      calculateInflowRate(idx, context.simulator());

```

Comment:
Sorry for not getting back to this sooner. It is just so occupied. 

some digging into the current PR, for 2D_OW_CTAQUIFER.DATA running, after the function calculateInflowRate(), Qai_ is NaN.